gateway/gateway.go

```diff
@@ -693,6 +701,66 @@ func (gateway *HandleT) getJobDataFromRequest(req *webRequestT) (jobData *jobFro
 	return
 }
 
+func setBatch(body []byte, reqType string) (batch []byte, err error) {
+	body, err = sjson.SetBytes(body, "type", reqType)
+	if err != nil {
+		err = errors.New(response.NotRudderEvent)
+		return
+	}
+	batch, err = sjson.SetRawBytes(BatchEvent, "batch.0", body)
+	if err != nil {
+		err = errors.New(response.NotRudderEvent)
+		return
+	}
+	return
+}
+
+func isNonIdentifiable(anonIDFromReq, userIDFromReq string) bool {
+	if anonIDFromReq == "" && userIDFromReq == "" {
+		return !allowReqsWithoutUserIDAndAnonymousID
+	}
+	return false
+}
+
+func buildUserID(userIDHeader, anonIDFromReq, userIDFromReq string) string {
+	if anonIDFromReq != "" {
+		return userIDHeader + DELIMITER + anonIDFromReq + DELIMITER + userIDFromReq
+	}
+	return userIDHeader + DELIMITER + userIDFromReq + DELIMITER + userIDFromReq
+}
+
+func (gateway *HandleT) isUserSuppressed(workspaceID, userID, sourceID string) bool {
+	if !enableSuppressUserFeature || gateway.suppressUserHandler == nil {
+		return false
+	}
+	return gateway.suppressUserHandler.IsSuppressedUser(
+		workspaceID,
+		userID,
+		sourceID,
+	)
+}
+
+// checks for the presence of messageId in the event
+//
+// sets to a new uuid if not present
+func checkMessageID(event map[string]interface{}) {
+	messageID := misc.MapLookup(event, "messageId")
```

Comment:
I'm not sure this `MapLookup` makes sense. It's used in many places but there are no tests for it. Do people really understand what it is supposed to do when multiple keys are passed to it and that recursion only happens if multiple keys are provided?

```go
func TestMapLookup(t *testing.T) {
	m := map[string]interface{}{
		"foo": "bar",
		"baz": "qux",
	}
	t.Log(MapLookup(m, "foo", "baz")) // prints <nil>

	m = map[string]interface{}{
		"foo": map[string]interface{}{
			"baz": "qux",
		},
	}
	t.Log(MapLookup(m, "foo", "baz")) // prints qux

	m = map[string]interface{}{
		"hello": map[string]interface{}{
			"foo": "bar",
		},
	}
	t.Log(MapLookup(m, "foo")) // prints <nil>
}
```

Anyhow, if that is how the function is supposed to work then using it gives you no benefits since it looks for the `messageId` key in the first level of the map only. You might as well just do:

```go
func checkMessageID(event map[string]interface{}) {
	messageID, _ := event["messageId"].(string)
	if strings.TrimSpace(messageID) == "" {
		event["messageId"] = uuid.New().String()
	}
}
```

And then you can remove `isEmptyString`. Let me know if I'm not making sense.

Reply:
True👍
makes sense

Reply:
I was worried about `messageID, _ := event["messageId"].(string)` part and didn't go this way initially.
This clarified things up for me.
Thanks!